staging/src/k8s.io/apiserver/pkg/storageversion/updater.go

```diff
@@ -35,23 +35,96 @@ type Client interface {
+		Type:               v1alpha1.AllEncodingVersionsEqual,
+		Status:             v1alpha1.ConditionFalse,
+		ObservedGeneration: sv.Generation,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		Reason:             "CommonEncodingVersionUnset",
+		Message:            "Common encoding version unset",
+	}
+	if sv.Status.CommonEncodingVersion != nil {
+		condition.Status = v1alpha1.ConditionTrue
+		condition.Reason = "CommonEncodingVersionSet"
+		condition.Message = "Common encoding version set"
+	}
+	forceTransition := false
+	if oldCommonEncodingVersion != nil && sv.Status.CommonEncodingVersion != nil &&
+		*oldCommonEncodingVersion != *sv.Status.CommonEncodingVersion {
+		forceTransition = true
+	}
+	setStatusCondition(&sv.Status.Conditions, condition, forceTransition)
+}
+
+func findStatusCondition(conditions []v1alpha1.StorageVersionCondition,
+	conditionType v1alpha1.StorageVersionConditionType) *v1alpha1.StorageVersionCondition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+// setStatusCondition sets the corresponding condition in conditions to newCondition.
+// conditions must be non-nil.
+// 1. if the condition of the specified type already exists (all fields of the existing condition are updated to
+//    newCondition, LastTransitionTime is set to now if the new status differs from the old status)
+// 2. if a condition of the specified type does not exist (LastTransitionTime is set to now() if unset, and newCondition is appended)
+// NOTE: forceTransition allows overwriting LastTransitionTime even when the status doesn't change.
+func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondition v1alpha1.StorageVersionCondition,
+	forceTransition bool) {
+	if conditions == nil {
 		return
 	}
-	firstVersion := sv.Status.StorageVersions[0].EncodingVersion
-	agreed := true
-	for _, ssv := range sv.Status.StorageVersions {
-		if ssv.EncodingVersion != firstVersion {
-			agreed = false
-			break
+	existingCondition := findStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		if newCondition.LastTransitionTime.IsZero() {
+			newCondition.LastTransitionTime = metav1.NewTime(time.Now())
 		}
+		*conditions = append(*conditions, newCondition)
+		return
 	}
-	if agreed {
-		sv.Status.CommonEncodingVersion = &firstVersion
-	} else {
-		sv.Status.CommonEncodingVersion = nil
+
+	if existingCondition.Status != newCondition.Status || forceTransition {
+		existingCondition.Status = newCondition.Status
+		if !newCondition.LastTransitionTime.IsZero() {
+			existingCondition.LastTransitionTime = newCondition.LastTransitionTime
+		} else {
+			existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		}
 	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
+	existingCondition.ObservedGeneration = newCondition.ObservedGeneration
 }
```

Comment:
I simplified the function a little, and it still passed the unit test, how about:

```golang
 func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondition v1alpha1.StorageVersionCondition,
     versionChanged bool) {
     if conditions == nil {
         return
     }
     existingCondition := findStatusCondition(*conditions, newCondition.Type)
     if existingCondition == nil {
         *conditions = append(*conditions, newCondition)
         return
     }
     statusChanged := existingCondition.Status != newCondition.Status
     if statusChanged || versionChanged {
         existingCondition.LastTransitionTime = newCondition.LastTransitionTime
     }
     existingCondition.Status = newCondition.Status
     existingCondition.Reason = newCondition.Reason
     existingCondition.Message = newCondition.Message
     existingCondition.ObservedGeneration = newCondition.ObservedGeneration
 }
```

Reply:
see https://github.com/kubernetes/kubernetes/pull/96825#discussion_r529899021

Reply:
Got it. Then how about

```golang
 func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondition v1alpha1.StorageVersionCondition,
     forceTransition bool) {
     if conditions == nil {
         return
     }
     if newCondition.LastTransitionTime.IsZero() {
          newCondition.LastTransitionTime = metav1.NewTime(time.Now())
     }
     existingCondition := findStatusCondition(*conditions, newCondition.Type)
     if existingCondition == nil {
         *conditions = append(*conditions, newCondition)
         return
     }
     statusChanged := existingCondition.Status != newCondition.Status
     if statusChanged || forceTransition {
         existingCondition.LastTransitionTime = newCondition.LastTransitionTime
     }
     existingCondition.Status = newCondition.Status
     existingCondition.Reason = newCondition.Reason
     existingCondition.Message = newCondition.Message
     existingCondition.ObservedGeneration = newCondition.ObservedGeneration
 }
```

Reply:
updated